### ISSUE TITLE
Duplicated rows when using fetchmany_arrow method

### ISSUE THREAD

**Issue by morleytj** [databricks/databricks-sql-python#286]
Hello, I've been using this package to automate some SQL pulldowns of a fairly large dataset, but have realized after running it that the fetchmany_arrow() method is potentially overlapping its returned results. I have included details as to the code I am running and the results below.

### MWE
```python
from databricks import sql
import os
from datetime import date
import sys
import pyarrow as pa
from pyarrow import csv

def stream_cursor_to_file_pyarrow(cursor,sql,filepath,size):
    cursor.execute(sql)
    #initial batch to get schema for writer
    res = cursor.fetchmany_arrow(size)
    with csv.CSVWriter(filepath, res.schema) as writer:
        writer.write_table(res)
        while True:
            res = cursor.fetchmany_arrow(size)
            if res.num_rows<1:
                break
            else:
                writer.write_table(res)

if __name__=="__main__":
    output_path = sys.argv[1]

    with sql.connect(server_hostname = os.getenv("DATABRICKS_SERVER_HOSTNAME"),
            http_path=os.getenv("DATABRICKS_HTTP_PATH"),
            access_token=os.getenv("DATABRICKS_TOKEN")) as connection:
        with connection.cursor() as cursor:
            print('connection established')
            q1_sql = "select * from table_name;"
            stream_cursor_to_file_pyarrow(cursor, q1_sql, output_path+'_counts_'+str(date.today())+'.csv',10000)
```

### Example data format
The data is in a long format in the table, and looks like this

| ID  | feature_name | feature_occurence | earliest_date_of_occurrence |
| ------------- | ------------- | ------------- | ------------- |
| A1  | F9  | 23  | 2010-04-04  | 
| A1  | F10  | 12  | 2009-11-20  | 
| B3  | CC1  | 2  | 2000-01-02  | 
| C4  | F9  | 34  | 2002-04-02  | 

### Error
The resulting file written to disk has duplicated rows in it (including the header), similar to the following example:
| ID  | feature_name | feature_occurence | earliest_date_of_occurrence |
| ------------- | ------------- | ------------- | ------------- |
| ID  | feature_name | feature_occurence | earliest_date_of_occurrence |
| A1  | F9  | 23  | 2010-04-04  | 
| **A1**  | **F10**  | **12**  | **2009-11-20**  | 
| **B3** | **CC1**  | **2**  | **2000-01-02**  | 
| C4  | F9  | 34  | 2002-04-02  | 
| ...  | ...  | ...  | ...  | 
| **B3**  | **CC1**  | **2**  | **2000-01-02**  | 
| X8  | CC9  | 3  | 2000-10-02  | 
| **A1**  | **F10**  | **12**  | **2009-11-20** | 

Interestingly the first duplicated row is directly at the 10000th index (same size I used for fetchmany_arrow). The first duplicated value is not the same as the first value in the first batch, meaning it is not a complete repasting of it, the duplicated row's first occurrence is at index=9216. I am wondering if this indicates some level of overlap in the fetch commands. This is also supported by the fact that all the following rows for a set number of rows are duplicated.

The total number of duplicates is 3849407, representing ~0.05% of the total number of records.

### Initial investigation of potential error sources:

I have checked to make sure the duplicates are not on the end of the SQL database, by running the following query:
```SQL
select * from tablename group by ID, feature_name, feature_occurence, earliest_date_of_occurrence having count(*)>1
```

However, this returns an empty set, indicating that the duplicates are being generated during retrieval or writing. The possibilities I am seeing as to the source of the error are either step of writing or the step of retrieval, but given the appearance of duplicated rows at the start of each batch I believe the error to be originated somewhere in the fetchmany_arrow() call.

Hopefully the error is not in my code somewhere, haha, and hopefully this is helpful in tracking down the potential issue.

Best,
Theodore

**Comment by susodapop** [comment]
Thanks for the fullsome write-up. I'm writing here to acknowledge that we've seen this and are working up a reproduction. Can you please share your version of Python, `pyarrow`, and `databricks-sql-connector`?

**Comment by morleytj** [comment]
Ah, I forgot to include those:
- python: version 3.10.9
- pyarrow: version 14.0.1
- databricks-sql-connector: version 3.0.0

Thanks for the quick reply! Let me know if there's any other information I can provide.

**Comment by Adomatic** [comment]
@susodapop Totally off-topic but you should probably look up 'fulsome' before you accidentally insult someone with it, like I did once in front of over 200 people. (-:

**Comment by susodapop** [comment]
Fair point, @Adomatic ;) Let's just say I'm trying to accelerate the revival of its positive connotations as discussed [here](https://www.merriam-webster.com/dictionary/fulsome):

> The senses shown above are the chief living senses of fulsome. Sense 2, which was a generalized term of disparagement in the late 17th century, is the least common of these. Fulsome became a point of dispute when sense 1, thought to be obsolete in the 19th century, began to be revived in the 20th. The dispute was exacerbated by the fact that the large dictionaries of the first half of the century missed the beginnings of the revival. Sense 1 has not only been revived but has spread in its application and continues to do so. The chief danger for the user of fulsome is ambiguity. Unless the context is made very clear, the reader or hearer cannot be sure whether such an expression as "fulsome praise" is meant in sense 1b or in sense 4.

**Comment by andrefurlan-db** [comment]
Hi, is this issue still occurring? I have not been able to reproduce it.

**Comment by morleytj** [comment]
Checking in to update that this issue is still occurring April 2024, and am curious if anyone has replicated it.

**Comment by morleytj** [comment]
Some extra context I noticed today -- generally there has very consistently been an extra header column generated as an indicator that rows were duplicated, one of the files I pulled today didn't have that extra column, so I checked and it didn't have any duplicates, though the other two did. This file is the smallest of the three I pulled, and had only two columns.

At first I thought this might be because of the batch size being larger than the retrieved table, but that isn't the case. Batch size was 10000 and row number of the nonduplicated table was 37,616. Unsure if relevant, but in the interest of providing all information, this unduplicated table is a table of two columns of unique ID's, and it's a select distinct of two of the columns from one of the two larger queries, one of which is duplicated.

**Comment by kravets-levko** [comment]
@morleytj can you please try to pass `use_cloud_fetch = false` to `sql.connect()` method and check if the behavior changes?

**Comment by morleytj** [comment]
@kravets-levko I just ran my pipeline with that argument in the `sql.connect()` method, and it actually seems to have worked completely, they con't seem to have any duplicates when that argument is included.

**Comment by kravets-levko** [comment]
Thank you @morleytj! This indicates that probably we have one more issue with CloudFetch feature, which is sad. But at least we have a direction. I'll ask you to test it a bit more with CloudFetch disabled - just to make sure it indeed helps. If you see duplicated rows again - please let me know

**Comment by morleytj** [comment]
Unfortunate, but thank you for the help in identifying the source of the error! The pipeline I'm using is run daily so I will keep you updated as to the output health.

**Comment by ion-elgreco** [comment]
@kravets-levko is this already resolved?

I was still using 2.9.6 and playing around with enabling cloud fetch, which gives improved performance, but if it returns an incorrect output I probably need to avoid it for now

**Comment by kravets-levko** [comment]
Hello here 👋 We just released v3.3.0 which includes a refactoring of CloudFetch-related code. Please give it a try and let me know if it helped with your issues or not (remember to enable CloudFetch via use_cloud_fetch=True). If you still see any issues - please enable debug logging (see https://github.com/databricks/databricks-sql-python/issues/383#issuecomment-2167627778) and share log output. Thank you!
